File: python/tests/test_media.py
"""Tests for rich media normalization — unified attachment model."""

from __future__ import annotations

import pytest

from unified_channel.media import (
    Attachment,
    MediaNormalizerMiddleware,
    MediaType,
    detect_media_type,
    normalize_attachment,
)
from unified_channel.types import (
    ContentType,
    Identity,
    MessageContent,
    UnifiedMessage,
)


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _make_msg(
    *,
    content_type: ContentType = ContentType.TEXT,
    text: str = "",
    media_url: str | None = None,
    media_type: str | None = None,
    raw=None,
    channel: str = "telegram",
) -> UnifiedMessage:
    return UnifiedMessage(
        id="msg-1",
        channel=channel,
        sender=Identity(id="user-1", username="tester"),
        content=MessageContent(
            type=content_type,
            text=text,
            media_url=media_url,
            media_type=media_type,
        ),
        raw=raw,
    )


# ---------------------------------------------------------------------------
# Attachment dataclass
# ---------------------------------------------------------------------------

def test_attachment_creation_all_fields():
    att = Attachment(
        type=MediaType.IMAGE,
        url="https://example.com/photo.jpg",
        data=b"\xff\xd8",
        filename="photo.jpg",
        mime_type="image/jpeg",
        size=12345,
        width=1920,
        height=1080,
        duration=None,
        thumbnail_url="https://example.com/thumb.jpg",
        metadata={"file_id": "abc123"},
    )
    assert att.type == MediaType.IMAGE
    assert att.url == "https://example.com/photo.jpg"
    assert att.data == b"\xff\xd8"
    assert att.filename == "photo.jpg"
    assert att.mime_type == "image/jpeg"
    assert att.size == 12345
    assert att.width == 1920
    assert att.height == 1080
    assert att.duration is None
    assert att.thumbnail_url == "https://example.com/thumb.jpg"
    assert att.metadata == {"file_id": "abc123"}


def test_attachment_defaults():
    att = Attachment(type=MediaType.DOCUMENT)
    assert att.url is None
    assert att.data is None
    assert att.filename is None
    assert att.size is None
    assert att.metadata == {}


# ---------------------------------------------------------------------------
# MediaType enum
# ---------------------------------------------------------------------------

def test_media_type_values():
    assert MediaType.IMAGE == "image"
    assert MediaType.VIDEO == "video"
    assert MediaType.AUDIO == "audio"
    assert MediaType.VOICE == "voice"
    assert MediaType.DOCUMENT == "document"
    assert MediaType.STICKER == "sticker"
    assert MediaType.LOCATION == "location"
    assert MediaType.CONTACT == "contact"


def test_media_type_is_str():
    assert isinstance(MediaType.IMAGE, str)
    assert str(MediaType.IMAGE) == "MediaType.IMAGE" or MediaType.IMAGE.value == "image"
    assert MediaType.IMAGE.value == "image"


# ---------------------------------------------------------------------------
# detect_media_type
# ---------------------------------------------------------------------------

def test_detect_from_mime_image():
    assert detect_media_type(mime_type="image/jpeg") == MediaType.IMAGE


def test_detect_from_mime_video():
    assert detect_media_type(mime_type="video/mp4") == MediaType.VIDEO


def test_detect_from_mime_audio():
    assert detect_media_type(mime_type="audio/mpeg") == MediaType.AUDIO


def test_detect_from_filename_png():
    assert detect_media_type(filename="photo.png") == MediaType.IMAGE


def test_detect_from_filename_mp4():
    assert detect_media_type(filename="clip.mp4") == MediaType.VIDEO


def test_detect_from_url_extension():
    assert detect_media_type(url="https://cdn.example.com/files/song.mp3?token=abc") == MediaType.AUDIO


def test_detect_from_url_image():
    assert detect_media_type(url="https://images.example.com/pic.webp") == MediaType.IMAGE


def test_detect_fallback_to_document():
    assert detect_media_type() == MediaType.DOCUMENT
    assert detect_media_type(mime_type="application/octet-stream") == MediaType.DOCUMENT


def test_detect_mime_takes_priority_over_filename():
    # MIME says video, filename says .mp3 — MIME wins
    result = detect_media_type(mime_type="video/webm", filename="audio.mp3")
    assert result == MediaType.VIDEO


def test_detect_filename_takes_priority_over_url():
    result = detect_media_type(filename="doc.pdf", url="https://example.com/image.png")
    assert result == MediaType.DOCUMENT


# ---------------------------------------------------------------------------
# normalize_attachment — Telegram style
# ---------------------------------------------------------------------------

def test_normalize_telegram_photo():
    raw = {
        "file_id": "AgACAgIAAxk",
        "mime_type": "image/jpeg",
        "file_size": 54321,
        "width": 800,
        "height": 600,
        "file_url": "https://api.telegram.org/file/bot123/photos/photo.jpg",
    }
    att = normalize_attachment(raw, "telegram")
    assert att.type == MediaType.IMAGE
    assert att.url == "https://api.telegram.org/file/bot123/photos/photo.jpg"
    assert att.mime_type == "image/jpeg"
    assert att.size == 54321
    assert att.width == 800
    assert att.height == 600
    assert att.metadata["file_id"] == "AgACAgIAAxk"


def test_normalize_telegram_voice():
    raw = {
        "type": "voice",
        "file_id": "AwACAgIAAxk",
        "mime_type": "audio/ogg",
        "file_size": 9876,
        "duration": 5.2,
        "file_url": "https://api.telegram.org/file/bot123/voice/msg.ogg",
    }
    att = normalize_attachment(raw, "telegram")
    assert att.type == MediaType.VOICE
    assert att.duration == 5.2
    assert att.metadata["file_id"] == "AwACAgIAAxk"


# ---------------------------------------------------------------------------
# normalize_attachment — Discord style
# ---------------------------------------------------------------------------

def test_normalize_discord_attachment():
    raw = {
        "url": "https://cdn.discordapp.com/attachments/123/456/image.png",
        "proxy_url": "https://media.discordapp.net/attachments/123/456/image.png",
        "content_type": "image/png",
        "filename": "image.png",
        "size": 102400,
        "width": 1024,
        "height": 768,
    }
    att = normalize_attachment(raw, "discord")
    assert att.type == MediaType.IMAGE
    assert att.url == "https://cdn.discordapp.com/attachments/123/456/image.png"
    assert att.filename == "image.png"
    assert att.mime_type == "image/png"
    assert att.width == 1024


# ---------------------------------------------------------------------------
# normalize_attachment — location / contact
# ---------------------------------------------------------------------------

def test_normalize_location():
    raw = {
        "type": "location",
        "latitude": 37.7749,
        "longitude": -122.4194,
    }
    att = normalize_attachment(raw, "telegram")
    assert att.type == MediaType.LOCATION
    assert att.metadata["latitude"] == 37.7749
    assert att.metadata["longitude"] == -122.4194


def test_normalize_contact():
    raw = {
        "type": "contact",
        "phone_number": "+1234567890",
        "first_name": "Alice",
        "last_name": "Smith",
    }
    att = normalize_attachment(raw, "telegram")
    assert att.type == MediaType.CONTACT
    assert att.metadata["phone_number"] == "+1234567890"
    assert att.metadata["first_name"] == "Alice"


def test_normalize_sticker():
    raw = {
        "type": "sticker",
        "file_id": "sticker123",
        "mime_type": "image/webp",
        "width": 512,
        "height": 512,
        "file_url": "https://api.telegram.org/file/bot123/stickers/s.webp",
    }
    att = normalize_attachment(raw, "telegram")
    assert att.type == MediaType.STICKER
    assert att.width == 512


# ---------------------------------------------------------------------------
# MediaNormalizerMiddleware
# ---------------------------------------------------------------------------

@pytest.mark.asyncio
async def test_middleware_adds_attachments_from_content():
    mw = MediaNormalizerMiddleware()
    msg = _make_msg(
        content_type=ContentType.MEDIA,
        media_url="https://example.com/photo.jpg",
        media_type="image/jpeg",
    )
    result = await mw.process(msg, _echo_handler)
    assert "attachments" in msg.metadata
    atts = msg.metadata["attachments"]
    assert len(atts) == 1
    assert atts[0].type == MediaType.IMAGE
    assert atts[0].url == "https://example.com/photo.jpg"


@pytest.mark.asyncio
async def test_middleware_adds_attachments_from_raw():
    mw = MediaNormalizerMiddleware()
    msg = _make_msg(
        raw={
            "attachments": [
                {
                    "url": "https://cdn.example.com/video.mp4",
                    "content_type": "video/mp4",
                    "size": 5000000,
                }
            ]
        }
    )
    result = await mw.process(msg, _echo_handler)
    assert "attachments" in msg.metadata
    assert msg.metadata["attachments"][0].type == MediaType.VIDEO


@pytest.mark.asyncio
async def test_middleware_passthrough_text_only():
    mw = MediaNormalizerMiddleware()
    msg = _make_msg(text="hello world")
    result = await mw.process(msg, _echo_handler)
    assert "attachments" not in msg.metadata
    assert result == "echo"


@pytest.mark.asyncio
async def test_middleware_skips_oversized_attachments():
    mw = MediaNormalizerMiddleware(max_size=1000)
    msg = _make_msg(
        raw={
            "attachments": [
                {
                    "url": "https://cdn.example.com/big.zip",
                    "content_type": "application/zip",
                    "size": 5000,
                },
                {
                    "url": "https://cdn.example.com/small.jpg",
                    "content_type": "image/jpeg",
                    "size": 500,
                },
            ]
        }
    )
    await mw.process(msg, _echo_handler)
    atts = msg.metadata["attachments"]
    assert len(atts) == 1
    assert atts[0].mime_type == "image/jpeg"


@pytest.mark.asyncio
async def test_middleware_combines_content_and_raw():
    mw = MediaNormalizerMiddleware()
    msg = _make_msg(
        content_type=ContentType.MEDIA,
        media_url="https://example.com/thumb.jpg",
        media_type="image/jpeg",
        raw={
            "attachments": [
                {"url": "https://example.com/doc.pdf", "content_type": "application/pdf"},
            ]
        },
    )
    await mw.process(msg, _echo_handler)
    assert len(msg.metadata["attachments"]) == 2


@pytest.mark.asyncio
async def test_voice_vs_audio_distinction():
    """Voice messages should remain VOICE, not collapse to AUDIO."""
    raw_voice = {"type": "voice", "mime_type": "audio/ogg", "duration": 3.0}
    att = normalize_attachment(raw_voice, "telegram")
    assert att.type == MediaType.VOICE

    raw_audio = {"mime_type": "audio/mpeg", "file_name": "song.mp3"}
    att2 = normalize_attachment(raw_audio, "telegram")
    assert att2.type == MediaType.AUDIO


# ---------------------------------------------------------------------------
# Handler helper
# ---------------------------------------------------------------------------

async def _echo_handler(msg: UnifiedMessage):
    return "echo"

File: rust/tests/test_media.rs
//! Tests for media types, detection, normalization, and middleware.

use std::collections::HashMap;
use unified_channel::media::*;
use unified_channel::middleware::*;
use unified_channel::types::*;

#[test]
fn test_detect_from_mime_image() {
    assert_eq!(detect_media_type(Some("image/png"), None, None), MediaType::Image);
    assert_eq!(detect_media_type(Some("image/jpeg"), None, None), MediaType::Image);
}

#[test]
fn test_detect_from_mime_video() {
    assert_eq!(detect_media_type(Some("video/mp4"), None, None), MediaType::Video);
}

#[test]
fn test_detect_from_mime_audio() {
    assert_eq!(detect_media_type(Some("audio/mpeg"), None, None), MediaType::Audio);
}

#[test]
fn test_detect_from_mime_sticker() {
    assert_eq!(
        detect_media_type(Some("application/x-tgsticker"), None, None),
        MediaType::Sticker
    );
}

#[test]
fn test_detect_from_filename() {
    assert_eq!(detect_media_type(None, Some("photo.jpg"), None), MediaType::Image);
    assert_eq!(detect_media_type(None, Some("video.mp4"), None), MediaType::Video);
    assert_eq!(detect_media_type(None, Some("song.mp3"), None), MediaType::Audio);
    assert_eq!(detect_media_type(None, Some("doc.pdf"), None), MediaType::Document);
}

#[test]
fn test_detect_from_url() {
    assert_eq!(
        detect_media_type(None, None, Some("https://example.com/img.png")),
        MediaType::Image
    );
    assert_eq!(
        detect_media_type(None, None, Some("https://cdn.example.com/video.webm?token=abc")),
        MediaType::Video
    );
}

#[test]
fn test_detect_fallback_document() {
    assert_eq!(detect_media_type(None, None, None), MediaType::Document);
    assert_eq!(
        detect_media_type(None, None, Some("https://example.com/file")),
        MediaType::Document
    );
}

#[test]
fn test_detect_mime_priority_over_filename() {
    // MIME says video, filename says image — MIME wins
    assert_eq!(
        detect_media_type(Some("video/mp4"), Some("photo.jpg"), None),
        MediaType::Video
    );
}

#[test]
fn test_attachment_new() {
    let att = Attachment::new(MediaType::Image);
    assert_eq!(att.media_type, MediaType::Image);
    assert!(att.url.is_none());
    assert!(att.data.is_none());
    assert!(att.filename.is_none());
}

#[test]
fn test_normalize_attachment_generic() {
    let mut raw = HashMap::new();
    raw.insert("url".to_string(), serde_json::json!("https://example.com/photo.jpg"));
    raw.insert("mime_type".to_string(), serde_json::json!("image/jpeg"));
    raw.insert("file_size".to_string(), serde_json::json!(12345));
    raw.insert("width".to_string(), serde_json::json!(800));
    raw.insert("height".to_string(), serde_json::json!(600));

    let att = normalize_attachment(&raw, "telegram");
    assert_eq!(att.media_type, MediaType::Image);
    assert_eq!(att.url.as_deref(), Some("https://example.com/photo.jpg"));
    assert_eq!(att.mime_type.as_deref(), Some("image/jpeg"));
    assert_eq!(att.size, Some(12345));
    assert_eq!(att.width, Some(800));
    assert_eq!(att.height, Some(600));
}

#[test]
fn test_normalize_attachment_location() {
    let mut raw = HashMap::new();
    raw.insert("type".to_string(), serde_json::json!("location"));
    raw.insert("latitude".to_string(), serde_json::json!(37.7749));
    raw.insert("longitude".to_string(), serde_json::json!(-122.4194));

    let att = normalize_attachment(&raw, "telegram");
    assert_eq!(att.media_type, MediaType::Location);
    assert!(att.metadata.contains_key("latitude"));
    assert!(att.metadata.contains_key("longitude"));
}

#[test]
fn test_normalize_attachment_contact() {
    let mut raw = HashMap::new();
    raw.insert("type".to_string(), serde_json::json!("contact"));
    raw.insert("phone_number".to_string(), serde_json::json!("+15551234"));
    raw.insert("first_name".to_string(), serde_json::json!("Alice"));

    let att = normalize_attachment(&raw, "telegram");
    assert_eq!(att.media_type, MediaType::Contact);
    assert!(att.metadata.contains_key("phone_number"));
    assert!(att.metadata.contains_key("first_name"));
}

#[test]
fn test_normalize_attachment_voice() {
    let mut raw = HashMap::new();
    raw.insert("type".to_string(), serde_json::json!("voice"));
    raw.insert("url".to_string(), serde_json::json!("https://example.com/voice.ogg"));
    raw.insert("duration".to_string(), serde_json::json!(5.5));
    raw.insert("file_id".to_string(), serde_json::json!("file123"));

    let att = normalize_attachment(&raw, "telegram");
    assert_eq!(att.media_type, MediaType::Voice);
    assert_eq!(att.duration, Some(5.5));
    assert_eq!(
        att.metadata.get("file_id"),
        Some(&serde_json::json!("file123"))
    );
}

#[test]
fn test_normalize_attachment_sticker() {
    let mut raw = HashMap::new();
    raw.insert("type".to_string(), serde_json::json!("sticker"));
    raw.insert("url".to_string(), serde_json::json!("https://example.com/sticker.webp"));
    raw.insert("width".to_string(), serde_json::json!(512));
    raw.insert("height".to_string(), serde_json::json!(512));

    let att = normalize_attachment(&raw, "telegram");
    assert_eq!(att.media_type, MediaType::Sticker);
    assert_eq!(att.width, Some(512));
}

#[tokio::test]
async fn test_media_normalizer_middleware_adds_attachments() {
    let mw = MediaNormalizerMiddleware::new();
    let handler: Handler = handler_fn(|msg: UnifiedMessage| async move {
        // Check that attachments were added to metadata
        if msg.metadata.contains_key("attachments") {
            HandlerResult::Text("has_attachments".into())
        } else {
            HandlerResult::Text("no_attachments".into())
        }
    });

    let mut msg = UnifiedMessage::new(
        "telegram",
        Identity::new("u1"),
        MessageContent::media("photo", "https://example.com/img.png", Some("image/png".into())),
    );

    let result = mw.process(msg, &handler).await;
    assert!(matches!(result, HandlerResult::Text(s) if s == "has_attachments"));
}

#[tokio::test]
async fn test_media_normalizer_no_media_passthrough() {
    let mw = MediaNormalizerMiddleware::new();
    let handler: Handler = handler_fn(|msg: UnifiedMessage| async move {
        if msg.metadata.contains_key("attachments") {
            HandlerResult::Text("has_attachments".into())
        } else {
            HandlerResult::Text("no_attachments".into())
        }
    });

    let msg = UnifiedMessage::new(
        "telegram",
        Identity::new("u1"),
        MessageContent::text("just text"),
    );

    let result = mw.process(msg, &handler).await;
    assert!(matches!(result, HandlerResult::Text(s) if s == "no_attachments"));
}

#[test]
fn test_attachment_serialize_roundtrip() {
    let mut att = Attachment::new(MediaType::Image);
    att.url = Some("https://example.com/img.png".into());
    att.mime_type = Some("image/png".into());
    att.size = Some(5000);

    let json = serde_json::to_string(&att).unwrap();
    let deserialized: Attachment = serde_json::from_str(&json).unwrap();
    assert_eq!(deserialized.media_type, MediaType::Image);
    assert_eq!(deserialized.url.as_deref(), Some("https://example.com/img.png"));
    assert_eq!(deserialized.size, Some(5000));
}
